osm2streets/src/transform/classify_intersections.rs

```diff
@@ -1,85 +1,229 @@
+        .iter()
+        .map(|(s, d)| (roads[*s].id, roads[*d].id))
+        .collect();
+    match worst_conflict {
+        Cross => (Crossing, Cross, full_connections),
+        c => (
+            if roads.len() == 2 {
+                Connection
+            } else {
+                MultiConnection
+            },
+            c,
+            full_connections,
+        ),
+    }
+}
+
+fn can_drive_out_of(road: &Road, which_end: NodeID) -> bool {
+    if let Some(driving_dir) = road.oneway_for_driving() {
+        let required_dir = if road.id.i2 == which_end {
+            Direction::Fwd
+        } else {
+            Direction::Back
+        };
+        return driving_dir == required_dir;
+    }
+    return true;
+}
+
+fn can_drive_into(road: &Road, which_end: NodeID) -> bool {
+    if let Some(driving_dir) = road.oneway_for_driving() {
+        let required_dir = if road.id.i1 == which_end {
+            Direction::Fwd
+        } else {
+            Direction::Back
+        };
+        return driving_dir == required_dir;
+    }
+    return true;
+}
 
-        // detect the simple case of a single road splitting
-        if num_roads_inout == 1 {
-            // TODO Determine if it is possible to turn from the 'in' road to the 'out' road.
-            // If not, then we have a dual carriageway split.
-            if false {
-                return MultiConnection;
+fn turn_is_allowed(src: &Road, dst: &Road) -> bool {
+    let mut has_exclusive_allows = false;
+    for (t, other) in src.turn_restrictions.iter() {
+        match t {
+            RestrictionType::BanTurns => {
+                if *other == dst.id {
+                    return false;
+                }
+            }
+            RestrictionType::OnlyAllowTurns => {
+                if *other == dst.id {
+                    return true;
+                }
+                has_exclusive_allows = true;
             }
         }
     }
+    !has_exclusive_allows
+}
+
+fn calc_conflict(a: &(usize, usize), b: &(usize, usize), side: DrivingSide) -> ConflictType {
+    use ConflictType::*;
+
+    // If the traffic starts of ends at the same place in the same direction...
+    if a.0 == b.0 && a.1 == b.1 {
+        return Uncontested;
+    }
+    if a.0 == b.0 {
+        return Diverge;
+    }
+    if a.1 == b.1 {
+        return Merge;
+    }
 
-    // A Merge is characterised by an uninterrupted road (that would qualify as a Connection or
-    // MultiConnection), with additional connected roads that yield.
-    // TODO Combine roads into corridors and count them.
+    // The intersection has a boundary that we have labelled 0 to n-1 in clockwise order (from an
+    // arbitrary point), like a string laying in a circle. If we represent `a` as an arc from one
+    // point on the string to another, then there is a section of the string between the two points,
+    // connecting them the two points and two ends of string "on the outside". A second arc, `b`,
```

Comment:
```suggestion
    // connecting them and two ends of string "on the outside". A second arc, `b`,
```

(I'm nitpicking here because I'm reading stuff super carefully to make sure I understand! The diagrams make it easy to follow along)